### LSE Data Analytics Online Career Accelerator

# DA201: Data Analytics using Python

## Assignment template: COVID-19 data

## Student Note
This template can be used to better understand the suggested assignment workflow and how to approach the questions. You are welcome to add code and Markdown blocks to the various sections to add either code or comments. Make sure to add code cells as applicable, and to comment all your code blocks.

You have the option to populate your Notebook with all the elements typically contained within the report, or to submit a separate report. In the case of submitting your Notebook, you can embed images, links and text where appropriate in addition to the text notes added.

**SPECIAL NOTE**
- Submit your Jupyter Notebook with the following naming convention: `LSE_DA201_assignment_[your name]_[your surname]` (remove the square brackets).
- You should submit a zipped folder containing all the elements used in your Notebook (data files, images, etc.).

> ***Markdown notes:*** Remember to change cell types to `Markdown`. You can review [Markdown basics](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax) to find out how to add formatted text, links and images to your Markdown documents.

## 0) Environment preparation
These settings are provided for you. You do not need to make any changes.

In [2]:
# Import the required libraries and set the plotting options
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

sns.set(rc = {'figure.figsize':(15,10)})

### Required: Report submission:
Insert URL (to your public GitHub repository) and a screenshot - double click cell to edit
- [My Github Repo](https://github.com/username/reponame)
- Screenshot demo (replace with your own).

!['My Github screenshot](http://github.com/apvoges/lse-ca/blob/main/GitHubScreenshot.png?raw=true)
(Note that this only works if your repo is set to **public**. Alternatively you need to refer to a local image and include this image in your submission.)

### 1.2) Presentation expectations:
- Describe the role and how workflow tools such as GitHub can be used to add value to organisations.

### Optional for notebook/Required for presentation.
- You can use this cell as placeholder for bullet points to include in your presentation. 
- This section will not be graded in the Notebook, grades awarded based on presentation content only. 

(Double-click to edit)

# Assignment 2

## 2) Assignment activity 2: 

### 2.1) Report expectations:
- Load the files `covid_19_uk_cases.csv` and `covid_19_uk_vaccinated.csv` and explore the data.
- Explore the data using the `info()`, `describe()`, `shape` and `value_counts()` methods, and note the observations regarding data types, number or records and features.
- Identify missing data.
- Filter/subset data.
- Aggregate data (totals and by month).
- Note observations.

### Loading CSVs in DataFrames

In [4]:
cov = pd.read_csv('covid_19_uk_cases.csv')
vac = pd.read_csv('covid_19_uk_vaccinated.csv')

### Exploring loaded DataFrames

##### First and last five rows of Covid and Vaccinated DataFrames for verification 

In [7]:
cov.head()

Province/State  Country/Region      Lat     Long ISO 3166-1 Alpha 3-Codes  \
0       Anguilla  United Kingdom  18.2206 -63.0686                      AIA   
1       Anguilla  United Kingdom  18.2206 -63.0686                      AIA   
2       Anguilla  United Kingdom  18.2206 -63.0686                      AIA   
3       Anguilla  United Kingdom  18.2206 -63.0686                      AIA   
4       Anguilla  United Kingdom  18.2206 -63.0686                      AIA   

                   Sub-region Name  Intermediate Region Code        Date  \
0  Latin America and the Caribbean                        29  2020-01-22   
1  Latin America and the Caribbean                        29  2020-01-23   
2  Latin America and the Caribbean                        29  2020-01-24   
3  Latin America and the Caribbean                        29  2020-01-25   
4  Latin America and the Caribbean                        29  2020-01-26   

   Deaths  Cases  Recovered  Hospitalised  
0     0.0    0.0        0.0           0.0  
1     0.0    0.0        0.0           0.0  
2     0.0    0.0        0.0           0.0  
3     0.0    0.0        0.0           0.0  
4     0.0    0.0        0.0           0.0

In [12]:
cov.tail()

Province/State  Country/Region      Lat   Long ISO 3166-1 Alpha 3-Codes  \
7579         Others  United Kingdom  55.3781 -3.436                      GBR   
7580         Others  United Kingdom  55.3781 -3.436                      GBR   
7581         Others  United Kingdom  55.3781 -3.436                      GBR   
7582         Others  United Kingdom  55.3781 -3.436                      GBR   
7583         Others  United Kingdom  55.3781 -3.436                      GBR   

      Sub-region Name  Intermediate Region Code        Date    Deaths  \
7579  Northern Europe                         0  2021-10-10  137735.0   
7580  Northern Europe                         0  2021-10-11  137763.0   
7581  Northern Europe                         0  2021-10-12  137944.0   
7582  Northern Europe                         0  2021-10-13  138080.0   
7583  Northern Europe                         0  2021-10-14  138237.0   

          Cases  Recovered  Hospitalised  
7579  8154306.0        0.0         378.0  
7580  8193769.0        0.0         386.0  
7581  8231437.0        0.0         386.0  
7582  8272883.0        0.0           0.0  
7583  8317439.0        0.0           0.0

In [10]:
vac.head()

Province/State  Country/Region      Lat     Long ISO 3166-1 Alpha 3-Codes  \
0       Anguilla  United Kingdom  18.2206 -63.0686                      AIA   
1       Anguilla  United Kingdom  18.2206 -63.0686                      AIA   
2       Anguilla  United Kingdom  18.2206 -63.0686                      AIA   
3       Anguilla  United Kingdom  18.2206 -63.0686                      AIA   
4       Anguilla  United Kingdom  18.2206 -63.0686                      AIA   

                   Sub-region Name  Intermediate Region Code        Date  \
0  Latin America and the Caribbean                        29  2020-01-22   
1  Latin America and the Caribbean                        29  2020-01-23   
2  Latin America and the Caribbean                        29  2020-01-24   
3  Latin America and the Caribbean                        29  2020-01-25   
4  Latin America and the Caribbean                        29  2020-01-26   

   Vaccinated  First Dose  Second Dose  
0           0           0            0  
1           0           0            0  
2           0           0            0  
3           0           0            0  
4           0           0            0

In [11]:
vac.tail()

Province/State  Country/Region      Lat   Long ISO 3166-1 Alpha 3-Codes  \
7579         Others  United Kingdom  55.3781 -3.436                      GBR   
7580         Others  United Kingdom  55.3781 -3.436                      GBR   
7581         Others  United Kingdom  55.3781 -3.436                      GBR   
7582         Others  United Kingdom  55.3781 -3.436                      GBR   
7583         Others  United Kingdom  55.3781 -3.436                      GBR   

      Sub-region Name  Intermediate Region Code        Date  Vaccinated  \
7579  Northern Europe                         0  2021-10-10        1070   
7580  Northern Europe                         0  2021-10-11        1300   
7581  Northern Europe                         0  2021-10-12        1482   
7582  Northern Europe                         0  2021-10-13           0   
7583  Northern Europe                         0  2021-10-14           0   

      First Dose  Second Dose  
7579        1216         1070  
7580        1604         1300  
7581        2027         1482  
7582           0            0  
7583           0            0

##### Determining number of ROWS and COLUMNS in 'COVID' and 'VACCINATED' DataFrames

In [5]:
print(cov.shape)

(7584, 12)


In [6]:
print(vac.shape)

(7584, 11)


##### Determining Data Types of 'COVID' and 'VACCINATED' DataFrames

In [13]:
print(vac.dtypes)

Province/State               object
Country/Region               object
Lat                         float64
Long                        float64
ISO 3166-1 Alpha 3-Codes     object
Sub-region Name              object
Intermediate Region Code      int64
Date                         object
Vaccinated                    int64
First Dose                    int64
Second Dose                   int64
dtype: object


In [14]:
print(cov.dtypes)

Province/State               object
Country/Region               object
Lat                         float64
Long                        float64
ISO 3166-1 Alpha 3-Codes     object
Sub-region Name              object
Intermediate Region Code      int64
Date                         object
Deaths                      float64
Cases                       float64
Recovered                   float64
Hospitalised                float64
dtype: object


In [15]:
print(vac.dtypes)

Province/State               object
Country/Region               object
Lat                         float64
Long                        float64
ISO 3166-1 Alpha 3-Codes     object
Sub-region Name              object
Intermediate Region Code      int64
Date                         object
Vaccinated                    int64
First Dose                    int64
Second Dose                   int64
dtype: object


In [28]:
cov.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7584 entries, 0 to 7583
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Province/State            7584 non-null   object 
 1   Country/Region            7584 non-null   object 
 2   Lat                       7584 non-null   float64
 3   Long                      7584 non-null   float64
 4   ISO 3166-1 Alpha 3-Codes  7584 non-null   object 
 5   Sub-region Name           7584 non-null   object 
 6   Intermediate Region Code  7584 non-null   int64  
 7   Date                      7584 non-null   object 
 8   Deaths                    7582 non-null   float64
 9   Cases                     7582 non-null   float64
 10  Recovered                 7582 non-null   float64
 11  Hospitalised              7582 non-null   float64
dtypes: float64(6), int64(1), object(5)
memory usage: 711.1+ KB


In [29]:
cov.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7584 entries, 0 to 7583
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Province/State            7584 non-null   object 
 1   Country/Region            7584 non-null   object 
 2   Lat                       7584 non-null   float64
 3   Long                      7584 non-null   float64
 4   ISO 3166-1 Alpha 3-Codes  7584 non-null   object 
 5   Sub-region Name           7584 non-null   object 
 6   Intermediate Region Code  7584 non-null   int64  
 7   Date                      7584 non-null   object 
 8   Deaths                    7582 non-null   float64
 9   Cases                     7582 non-null   float64
 10  Recovered                 7582 non-null   float64
 11  Hospitalised              7582 non-null   float64
dtypes: float64(6), int64(1), object(5)
memory usage: 711.1+ KB


In [30]:
cov.value_counts()

Province/State                                Country/Region  Lat         Long        ISO 3166-1 Alpha 3-Codes  Sub-region Name                  Intermediate Region Code  Date        Deaths    Cases      Recovered  Hospitalised
Anguilla                                      United Kingdom   18.220600  -63.068600  AIA                       Latin America and the Caribbean  29                        2020-01-22  0.0       0.0        0.0        0.0             1
Isle of Man                                   United Kingdom   54.236100  -4.548100   IMN                       Northern Europe                  0                         2021-10-11  54.0      8121.0     0.0        631.0           1
Montserrat                                    United Kingdom   16.742498  -62.187366  MSR                       Latin America and the Caribbean  29                        2020-01-30  0.0       0.0        0.0        0.0             1
                                                                         

##### Number of missing values in 'Covid' DataFrame

In [16]:
cov_na = cov[cov.isna().any(axis=1)]

cov_na.shape

(2, 12)

###### Atif's Comment: Two missing values in 12 Columns in 'Covid' DataFrame above

##### Number of missing values in 'Vaccinated' DataFrame

In [17]:
vac_na = vac[vac.isna().any(axis=1)]

vac_na.shape

(0, 11)

###### Atif's Comment: No missing values in 11 columns in 'Vaccinated' DataFrame above

##### Filtering Gibraltar

In [20]:
cov_gibraltar = cov[cov['Province/State']=='Gibraltar']

cov_gibraltar

Province/State  Country/Region      Lat    Long ISO 3166-1 Alpha 3-Codes  \
3792      Gibraltar  United Kingdom  36.1408 -5.3536                      GIB   
3793      Gibraltar  United Kingdom  36.1408 -5.3536                      GIB   
3794      Gibraltar  United Kingdom  36.1408 -5.3536                      GIB   
3795      Gibraltar  United Kingdom  36.1408 -5.3536                      GIB   
3796      Gibraltar  United Kingdom  36.1408 -5.3536                      GIB   
...             ...             ...      ...     ...                      ...   
4419      Gibraltar  United Kingdom  36.1408 -5.3536                      GIB   
4420      Gibraltar  United Kingdom  36.1408 -5.3536                      GIB   
4421      Gibraltar  United Kingdom  36.1408 -5.3536                      GIB   
4422      Gibraltar  United Kingdom  36.1408 -5.3536                      GIB   
4423      Gibraltar  United Kingdom  36.1408 -5.3536                      GIB   

      Sub-region Name  Intermediate Region Code        Date  Deaths   Cases  \
3792  Southern Europe                         0  2020-01-22     0.0     0.0   
3793  Southern Europe                         0  2020-01-23     0.0     0.0   
3794  Southern Europe                         0  2020-01-24     0.0     0.0   
3795  Southern Europe                         0  2020-01-25     0.0     0.0   
3796  Southern Europe                         0  2020-01-26     0.0     0.0   
...               ...                       ...         ...     ...     ...   
4419  Southern Europe                         0  2021-10-10    97.0  5626.0   
4420  Southern Europe                         0  2021-10-11    97.0  5655.0   
4421  Southern Europe                         0  2021-10-12    97.0  5682.0   
4422  Southern Europe                         0  2021-10-13    97.0  5707.0   
4423  Southern Europe                         0  2021-10-14    97.0  5727.0   

      Recovered  Hospitalised  
3792        0.0           0.0  
3793        0.0           0.0  
3794        0.0           0.0  
3795        0.0           0.0  
3796        0.0           0.0  
...         ...           ...  
4419        0.0         858.0  
4420        0.0         876.0  
4421        0.0         876.0  
4422        0.0           0.0  
4423        0.0           0.0  

[632 rows x 12 columns]

In [24]:
pd.set_option("display.max_rows", None)


##### Subsetting: Selecting specified columns (D= Death, C= Cases, R = Recovered, H = Hospitalized )

In [25]:
cov_gibraltar_DCRH = cov_gibraltar [["Deaths","Cases","Recovered","Hospitalised"]]


In [26]:
cov_gibraltar_DCRH.describe()

Deaths        Cases    Recovered  Hospitalised
count  632.000000   632.000000   632.000000    632.000000
mean    40.208861  2237.109177  1512.821203   1027.625000
std     45.332832  2136.268090  1817.096755   1145.681058
min      0.000000     0.000000     0.000000      0.000000
25%      0.000000   177.000000   109.500000    157.750000
50%      5.000000  1036.500000   323.500000    675.500000
75%     94.000000  4286.000000  4122.500000   1548.000000
max     97.000000  5727.000000  4670.000000   4907.000000

In [27]:
cov.describe()

Lat         Long  Intermediate Region Code         Deaths  \
count  7584.000000  7584.000000               7584.000000    7582.000000   
mean     21.840267   -41.440047                 12.500000    6210.202321   
std      28.305518    30.527529                 14.009849   24860.790148   
min     -51.796300   -81.254600                  0.000000       0.000000   
25%      17.851075   -64.667625                  0.000000       0.000000   
50%      20.503650   -60.855483                  2.500000       1.000000   
75%      39.448675    -5.152225                 29.000000      25.000000   
max      55.378100    -2.364400                 29.000000  138237.000000   

              Cases    Recovered  Hospitalised  
count  7.582000e+03  7582.000000   7582.000000  
mean   2.147082e+05   454.691242    685.228568  
std    1.007557e+06  1010.642337    819.256635  
min    0.000000e+00     0.000000      0.000000  
25%    1.200000e+01     0.000000    105.000000  
50%    1.685000e+02    13.000000    390.000000  
75%    1.585000e+03   312.000000    955.750000  
max    8.317439e+06  8322.000000   4907.000000

In [ ]:
# Explore behaviour over time

In [1]:
# Explore and note observations
# Are there any visualisations that could be added here to make it easier?

In [ ]:
# We shall start with analysing the entire DataFrame of covid cases. 
# Data is collected against twelve terrorities of United Kingdom, 
# outside mainland, for a period of 22 Months or 632 days for each territory. 

### 2.2) Presentation expectations:
Use the process of exploring the data for Gibraltar as an example to provide a brief description of the various phases to help your team to understand the process. Keep it high level and make sure to focus on both specifics relating to the case (first dose, second dose per region, total and over time) and brief observations regarding the process. Assignment activity 2 considers basic data exploration.
- Can we make decisions based on total numbers only, or do trends over time offer additional insights?
- Why it is important to explore the data, what are the typical mistakes made in this phase?

## 3) Assignment activity 3: 

### 3.1) Report expectations:
- Merge and explore the data.
- Convert the data type of the Date column from object to DateTime.
- Create a data set that meets the expected parameters.
- Add calculated features to DataFrames (difference between first and second dose vaccinations).
- Filter and sort output.
- Observe totals and percentages as a total and over time.
- Note observations.

Merge the DataFrames without duplicating columns. The new DataFrame (e.g. `covid`) will have `7584` rows and the following columns: `Province/State, Country/Region, Date, Vaccinated, First Dose, Second Dose, Deaths, Cases, Recovered, Hospitalised`.

In [ ]:
# Join the DataFrames as covid where you merge cov and vac

In [ ]:
# Explore the new DataFrame

In [ ]:
# Fix the date column data type

In [ ]:
# Clean up / drop unnecessary columns 

In [ ]:
# Groupby and calculate difference between first and second dose

In [2]:
# Groupby and calculate the difference between first and second dose over time

### 3.2) Presentation expectations:
We use similar calculations and representations as we had in assignment activity 2, but now expand to look at all provinces. Assignment activity 3 is concerned with exploring data in the context of a specific business question (as opposed to general exploration in assignment activity 2).
- What insights can be gained from the data? (Description of all regions, assumptions and concerns, trends or patterns you have observed.)
- Are there limitations or assumptions that needs to be considered?
- Make sure to provide a brief overview of the data and typical considerations at this phase of analysis.

## 4) Assignment activity 4: 

The government is looking to promote second dose vaccinations and would like to know the best possible area to test a new campaign. They are looking for the highest number of people who have received a first dose and not a second dose. 
- Where should they target?
- Which provinces have the highest number (actual numbers) and highest relative numbers (second dose only/first dose)?
- Visualise both outputs.

### 4.1) Report expectations:
- Consider additional features (deaths and recoveries).
- Visualise the data.
- Note observations:
 - Do deaths follow the same patterns observed in vaccination data (daily vs cumulative)?
 - Do we need to separate groups of data for specific variables and analyse them in isolation (Others) to be able to observe the patterns?

### 4.2) Presentation expectations:
- What insights can be gained from the data?
- Why do we need to consider other features?
- **Hints**: 
 - Evalute different features to improve decision making (deaths and recoveries). 
 - Why it is important to explore data and use different views?
 - Highlight two or three suggestions to get junior team members started in terms of good practices.

In [ ]:
# Absolute numbers

In [ ]:
# Relative numbers (%)

In [ ]:
# Sort and display

In [ ]:
# Visualise

In [ ]:
# Smooth out the data by looking at monthly figures

In [ ]:
# Other features evaluated (data preparation, output and plots)

***Notes and observations:***
Your observations here. (Double click to edit)

***Examples could include:***
- Are there other trends in terms of recoveries or hospitalisations compared to other features that you found interesting and that may add value in terms of the decision making process?
- Any other observations regarding the data?
- Any suggestions for improvements and further analysis?
- What would your future data requirements be?

## 5) Assignment activity 5: External data 
In the next section, you were supplied with a sample file and the question was asked to determine whether there are additional `#tags` or keywords that could potentially provide insights into your COVID-19 analysis. While the sample set is limited, you were asked to review the provided file and demonstrate the typical steps and make recommendations regarding future use of similar data sets to provide richer insights.

### 5.1) Report expectations:
- Demonstrate basic ability to work with Twitter data.
- Search for hash-tags or keywords.
- Create DataFrames and visualisations.
- Note your observations.

In [ ]:
# Import the tweet data set


In [ ]:
# Explore the data: info(), head()


In [ ]:
# Explore the structure, count the tweets, get the elements of interest


In [ ]:
# Create a DataFrame with the text only


In [ ]:
# Loop through the messages and build a list of values containing the #-symbol


In [ ]:
# Filter and sort


In [ ]:
# Plot


### 5.2) Presentation expectations:
Discuss whether external data could potentially be used and whether it is a viable solution to pursue. Discuss your assumptions and suggestions. 

Points to consider:
- What insights can be gained from the data?
- What are the advantages and disadvantages of using external data?
- How would you suggest using external data in the project?

## 6) Assignment activity 6: 

### 6.1) Report expectations:
- Demonstrate using external function and interpret results.
- Note observations.

In [ ]:
# You can copy and paste the relevant code cells from the provided template here.

### 6.2) Presentation expectations:
- **Question 1**: We have heard of both qualitative and quantitative data from the previous consultant. What are the differences between the two? Should we use only one or both of these types of data and why? How can these be used in business predictions? Could you provide examples of each?
- **Question 2**: We have also heard a bit about the need for continuous improvement. Why should this be implemented, it seems like a waste of time. Why can’t we just implement the current project as it stands and move on to other pressing matters?
- **Question 3**: As a government, we adhere to all data protection requirements and have good governance in place. We only work with aggregated data and therefore will not expose any personal details. Have we covered everything from a data ethics standpoint? Is there anything else we need to implement from a data ethics perspective? 